src/monodroid/jni/embedded-assemblies.cc

```diff
@@ -256,312 +350,580 @@ EmbeddedAssemblies::add_type_mapping (TypeMappingInfo **info, const char *source
-	// System.dll, System.dll.mdb case
-	if (strncmp (assembly->name, entry_basename, strlen (assembly->name)) != 0) {
-		// That failed; try for System.dll, System.pdb case
-		const char *eb_ext  = strrchr (entry_basename, '.');
-		if (eb_ext == nullptr)
-			return false;
-		off_t basename_len    = static_cast<off_t>(eb_ext - entry_basename);
-		assert (basename_len > 0 && "basename must have a length!");
-		if (strncmp (assembly->name, entry_basename, static_cast<size_t>(basename_len)) != 0)
-			return false;
+	if (bundled_assemblies_size - bundled_assemblies_count <= BUNDLED_ASSEMBLIES_EXCESS_ITEMS_LIMIT) {
+		return;
 	}
 
-	mono_register_symfile_for_assembly (assembly->name, debug_contents, debug_size);
-
-	return true;
+	resize_bundled_assemblies (bundled_assemblies_count);
 }
 
 bool
-EmbeddedAssemblies::gather_bundled_assemblies_from_apk (const char* apk, monodroid_should_register should_register)
+EmbeddedAssemblies::zip_load_entries (int fd, const char *apk_name, size_t total_apk_count, monodroid_should_register should_register)
 {
-	int fd;
-	unzFile file;
-
-	zlib_filefunc_def funcs = {
-		md_mmap_open_file,  // zopen_file,
-		md_mmap_read_file,  // zread_file,
-		nullptr,            // zwrite_file,
-		md_mmap_tell_file,  // ztell_file,
-		md_mmap_seek_file,  // zseek_file,
-		md_mmap_close_file, // zclose_file
-		md_mmap_error_file, // zerror_file
-		nullptr             // opaque
-	};
+	uint32_t cd_offset;
+	uint32_t cd_size;
+	uint16_t cd_entries;
 
-	if ((fd = open (apk, O_RDONLY)) < 0) {
-		log_error (LOG_DEFAULT, "ERROR: Unable to load application package %s.", apk);
+	if (!zip_read_cd_info (fd, cd_offset, cd_size, cd_entries)) {
+		log_fatal (LOG_DEFAULT,  "Failed to read the EOCD record from APK file %s", apk_name);
+		return false;
+	}
+#ifdef DEBUG
+	log_info (LOG_DEFAULT, "Central directory offset: %u", cd_offset);
+	log_info (LOG_DEFAULT, "Central directory size: %u", cd_size);
+	log_info (LOG_DEFAULT, "Central directory entries: %u", cd_entries);
+#endif
+	off_t result = ::lseek (fd, static_cast<off_t>(cd_offset), SEEK_SET);
+	if (result < 0) {
+		log_fatal (LOG_DEFAULT, "Failed to seek to central directory position in the APK file %s", apk_name);
 		return false;
 	}
 
-	funcs.opaque = &fd;
+	if (bundled_assemblies == nullptr) {
+		bundled_assemblies_count = 0;
+		bundled_assemblies_size = MULTIPLY_WITH_OVERFLOW_CHECK (size_t, cd_entries, total_apk_count);
+		bundled_assemblies = new XamarinBundledAssembly[bundled_assemblies_size];
+	} else if (bundled_assemblies_size - bundled_assemblies_count <= cd_entries) {
+		resize_bundled_assemblies (ADD_WITH_OVERFLOW_CHECK (size_t, bundled_assemblies_size, cd_entries << 1));
+	}
 
-	if ((file = unzOpen2 (nullptr, &funcs)) != nullptr) {
-		do {
-			unz_file_info info;
-			uLong offset;
-			unsigned int *psize;
-			char cur_entry_name [256];
-			MonoBundledAssembly *cur;
+	simple_pointer_guard<uint8_t>  buf (new uint8_t[cd_size]);
```

Comment:
My C++ is assuredly rusty, but...  Shouldn't this be `simple_pointer_guard<uint8_t[]>`?  Otherwise, I imagine that the destructor would use`std::is_array_v<uint8_t>`, which would be *false*, which would cause the `new[]` to correspond with `delete`, *not* `delete[]`.